### PR TITLE
[ML] Update PyTorch and build oneDNN in linux-x86_64 docker file 

### PR DESCRIPTION
**Pull request by davidkyle**
non-issue as it is an update to unreleased code.

### CODE REVIEW

**Comment by joshdevins** [comment]
Does this close https://github.com/elastic/ml-cpp/issues/1954 ?

**Comment by droberts195** [comment]
> Does this close #1954 ?

Not completely, because it's only doing Linux.  There will be separate PRs for macOS and Windows so that we're using PyTorch 1.9 on all platforms before feature freeze.  However, in this cycle of updates we're only going to use oneDNN on linux-x86_64.  For the other platforms we'll just build PyTorch 1.9 as similarly as possible to how we built PyTorch 1.8, so the upgrade process should be quicker.

**Comment by davidkyle** [comment]
I and not sure exactly which MKL libraries to include, the ones I have included were testing on a Cascade Lake CPU. I'm assuming which of the various `libmkl_avx*.so` libraries used depend on the CPU instruction set, same for `lib_vml_avx*.so`

This is the full directory listing for the MKL libraries installed by in the yum package

```
total 1.9G
-rwxr-xr-x 1 root root  50M Sep 17  2020 libmkl_avx.so
-rwxr-xr-x 1 root root  47M Sep 17  2020 libmkl_avx2.so
-rwxr-xr-x 1 root root  61M Sep 17  2020 libmkl_avx512.so
-rwxr-xr-x 1 root root  64M Sep 17  2020 libmkl_avx512_mic.so
-rw-r--r-- 1 root root 1.2M Sep 17  2020 libmkl_blacs_intelmpi_ilp64.a
-rwxr-xr-x 1 root root 488K Sep 17  2020 libmkl_blacs_intelmpi_ilp64.so
-rw-r--r-- 1 root root 722K Sep 17  2020 libmkl_blacs_intelmpi_lp64.a
-rwxr-xr-x 1 root root 295K Sep 17  2020 libmkl_blacs_intelmpi_lp64.so
-rw-r--r-- 1 root root 1.3M Sep 17  2020 libmkl_blacs_openmpi_ilp64.a
-rwxr-xr-x 1 root root 489K Sep 17  2020 libmkl_blacs_openmpi_ilp64.so
-rw-r--r-- 1 root root 741K Sep 17  2020 libmkl_blacs_openmpi_lp64.a
-rwxr-xr-x 1 root root 300K Sep 17  2020 libmkl_blacs_openmpi_lp64.so
-rw-r--r-- 1 root root 1.2M Sep 17  2020 libmkl_blacs_sgimpt_ilp64.a
-rwxr-xr-x 1 root root 485K Sep 17  2020 libmkl_blacs_sgimpt_ilp64.so
-rw-r--r-- 1 root root 721K Sep 17  2020 libmkl_blacs_sgimpt_lp64.a
-rwxr-xr-x 1 root root 296K Sep 17  2020 libmkl_blacs_sgimpt_lp64.so
-rw-r--r-- 1 root root 649K Sep 17  2020 libmkl_blas95_ilp64.a
-rw-r--r-- 1 root root 643K Sep 17  2020 libmkl_blas95_lp64.a
-rw-r--r-- 1 root root 213K Sep 17  2020 libmkl_cdft_core.a
-rwxr-xr-x 1 root root 166K Sep 17  2020 libmkl_cdft_core.so
-rw-r--r-- 1 root root 640M Sep 17  2020 libmkl_core.a
-rwxr-xr-x 1 root root  72M Sep 17  2020 libmkl_core.so
-rwxr-xr-x 1 root root  40M Sep 17  2020 libmkl_def.so
-rw-r--r-- 1 root root  27M Sep 17  2020 libmkl_gf_ilp64.a
-rwxr-xr-x 1 root root  12M Sep 17  2020 libmkl_gf_ilp64.so
-rw-r--r-- 1 root root  28M Sep 17  2020 libmkl_gf_lp64.a
-rwxr-xr-x 1 root root  13M Sep 17  2020 libmkl_gf_lp64.so
-rw-r--r-- 1 root root  42M Sep 17  2020 libmkl_gnu_thread.a
-rwxr-xr-x 1 root root  30M Sep 17  2020 libmkl_gnu_thread.so
-rw-r--r-- 1 root root  27M Sep 17  2020 libmkl_intel_ilp64.a
-rwxr-xr-x 1 root root  12M Sep 17  2020 libmkl_intel_ilp64.so
-rw-r--r-- 1 root root  28M Sep 17  2020 libmkl_intel_lp64.a
-rwxr-xr-x 1 root root  13M Sep 17  2020 libmkl_intel_lp64.so
-rw-r--r-- 1 root root  86M Sep 17  2020 libmkl_intel_thread.a
-rwxr-xr-x 1 root root  62M Sep 17  2020 libmkl_intel_thread.so
-rw-r--r-- 1 root root 7.1M Sep 17  2020 libmkl_lapack95_ilp64.a
-rw-r--r-- 1 root root 7.1M Sep 17  2020 libmkl_lapack95_lp64.a
-rwxr-xr-x 1 root root  46M Sep 17  2020 libmkl_mc.so
-rwxr-xr-x 1 root root  48M Sep 17  2020 libmkl_mc3.so
-rw-r--r-- 1 root root  54M Sep 17  2020 libmkl_pgi_thread.a
-rwxr-xr-x 1 root root  40M Sep 17  2020 libmkl_pgi_thread.so
-rwxr-xr-x 1 root root 6.9M Sep 17  2020 libmkl_rt.so
-rw-r--r-- 1 root root  12M Sep 17  2020 libmkl_scalapack_ilp64.a
-rwxr-xr-x 1 root root 7.4M Sep 17  2020 libmkl_scalapack_ilp64.so
-rw-r--r-- 1 root root  12M Sep 17  2020 libmkl_scalapack_lp64.a
-rwxr-xr-x 1 root root 7.4M Sep 17  2020 libmkl_scalapack_lp64.so
-rw-r--r-- 1 root root  37M Sep 17  2020 libmkl_sequential.a
-rwxr-xr-x 1 root root  28M Sep 17  2020 libmkl_sequential.so
-rw-r--r-- 1 root root  53M Sep 17  2020 libmkl_tbb_thread.a
-rwxr-xr-x 1 root root  37M Sep 17  2020 libmkl_tbb_thread.so
-rwxr-xr-x 1 root root  15M Sep 17  2020 libmkl_vml_avx.so
-rwxr-xr-x 1 root root  15M Sep 17  2020 libmkl_vml_avx2.so
-rwxr-xr-x 1 root root  14M Sep 17  2020 libmkl_vml_avx512.so
-rwxr-xr-x 1 root root  15M Sep 17  2020 libmkl_vml_avx512_mic.so
-rwxr-xr-x 1 root root 7.3M Sep 17  2020 libmkl_vml_cmpt.so
-rwxr-xr-x 1 root root 8.0M Sep 17  2020 libmkl_vml_def.so
-rwxr-xr-x 1 root root  14M Sep 17  2020 libmkl_vml_mc.so
-rwxr-xr-x 1 root root  14M Sep 17  2020 libmkl_vml_mc2.so
-rwxr-xr-x 1 root root  14M Sep 17  2020 libmkl_vml_mc3.so
drwxr-xr-x 4 root root 4.0K Sep  7 09:14 locale
```

**Comment by droberts195** [comment]
> I and not sure exactly which MKL libraries to include

I had a look at what the build available from PyTorch looks like (from `libtorch-cxx11-abi-shared-with-deps-1.9.0+cpu.zip`).

It seems that they statically link MKL: there are no MKL shared libraries at all in the bundle but an `nm -C libtorch_cpu.so` reveals masses of MKL functions whose code is inside `libtorch_cpu.so`.  So we could do the same thing.  Possibly that would reuse the knowledge of the PyTorch developers to link in everything required and no more.

If we go for dynamic linking instead then https://software.intel.com/content/www/us/en/develop/tools/oneapi/components/onemkl/link-line-advisor.html might be a useful tool.  We should see what it recommends for a variety of different machine types.

**Comment by droberts195** [comment]
We certainly don't need any of the libraries with `ilp64` in their names, as we're using `lp64` variable sizes.  And we certainly don't need any of the libraries with `95` in their names, as they're for Fortran.

**Comment by davidkyle** [comment]
Currently this PR uses a version of MKL from 2020. With the rebranding to oneAPI the MKL repository has moved, details of how to install the 2021 versions of the library with yum are [here](https://software.intel.com/content/www/us/en/develop/documentation/installation-guide-for-intel-oneapi-toolkits-linux/top/installation/install-using-package-managers/yum-dnf-zypper.html)

The install location has also changed to`/opt/intel/oneapi/mkl/` from `/opt/intel/mkl/` and unfortunately the libtorch `FindMKL.cmake` does not search for MKL in the new location failing the build (`python setup.py install`). 

cmake is infinitely configurable so no doubt there is a way to [specify a different location](https://github.com/pytorch/pytorch/blob/5be8a4e0278ab6e6d7c7ab053be3fbc0b9488f8b/cmake/Modules/FindMKL.cmake#L49) if we wish to explore that option. 

For future reference the docker file script to install the oneAPI MKL would be:
```
# Install Intel MKL
RUN \
  { tee /tmp/oneAPI.repo << EOF
  [oneAPI]
  name=Intel® oneAPI repository
  baseurl=https://yum.repos.intel.com/oneapi
  enabled=1
  gpgcheck=1
  repo_gpgcheck=1
  gpgkey=https://yum.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS.PUB
  EOF
  } && mv /tmp/oneAPI.repo /etc/yum.repos.d && \
  yum -y install intel-oneapi-mkl-2021.3.0.x86_64 && \
  cp /opt/intel/oneapi/mkl/2021.3.0/lib/intel64/libmkl_intel_lp64.so /usr/local/gcc93/lib && \
  cp /opt/intel/oneapi/mkl/2021.3.0/lib/intel64/libmkl_core.so /usr/local/gcc93/lib && \
  cp /opt/intel/oneapi/mkl/2021.3.0/lib/intel64/libmkl_def.so /usr/local/gcc93/lib && \
  cp /opt/intel/oneapi/mkl/2021.3.0/lib/intel64/libmkl_gnu_thread.so /usr/local/gcc93/lib && \
  cp /opt/intel/oneapi/mkl/2021.3.0/lib/intel64/libmkl_avx*.so /usr/local/gcc93/lib && \
  cp /opt/intel/oneapi/mkl/2021.3.0/lib/intel64/libmkl_vml*.so /usr/local/gcc93/lib
```

**Comment by davidkyle** [comment]
I tested this on an older Intel Xeon CPU that supports SSE2 but not AVX and saw considerable improvements in performance using MKL and oneDNN

File: build-setup/linux.md [hunk 1]
```diff
@@ -261,7 +261,30 @@ make
...
+Intel Maths Kernal Library is an optimized BLAS library which
+greatly improves the performance of PyTorch CPU inference 
+when Pytorch is built with it. 
+
+```
+yum-config-manager --add-repo https://yum.repos.intel.com/mkl/setup/intel-mkl.repo
+yum -y install intel-mkl-2020.4-912
+```
+
+Then copy the required libraries to the system dircectory
```
Comment (droberts195) [inline build-setup/linux.md:275]:
```suggestion
Then copy the required libraries to the system directory:
```

File: build-setup/linux.md [hunk 2]
```diff
@@ -261,7 +261,30 @@ make
 sudo make altinstall
 ```
 
-### PyTorch 1.8.0
+### Intel MKL
+
+Intel Maths Kernal Library is an optimized BLAS library which
+greatly improves the performance of PyTorch CPU inference 
+when Pytorch is built with it.
```
Comment (droberts195) [inline build-setup/linux.md:268]:
```suggestion
when PyTorch is built with it. 
```

**Review by droberts195 — APPROVED** [review]
LGTM

**Comment by davidkyle** [comment]
retest

`docker.elastic.co/ml-dev/ml-linux-build:18` is now published

**Comment by davidkyle** [comment]
retest

**Pull request merged**